src/main/java/net/imglib2/realtransform/DeformationFieldTransform.java

```diff
@@ -0,0 +1,150 @@
+ * 
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package net.imglib2.realtransform;
+
+import net.imglib2.RandomAccessibleInterval;
+import net.imglib2.RealLocalizable;
+import net.imglib2.RealPositionable;
+import net.imglib2.RealRandomAccess;
+import net.imglib2.RealRandomAccessible;
+import net.imglib2.interpolation.randomaccess.NLinearInterpolatorFactory;
+import net.imglib2.type.numeric.RealType;
+import net.imglib2.view.Views;
+
+/**
+ * An <em>n</em>-dimensional deformation field.
+ * <p> 
+ * Wraps a {@link RandomAccessibleInterval} of dimensionality D+1 ("def")
+ * and interprets it as a D-dimensional {@link RealTransform}.  The last
+ * dimension of of the {@link RandomAccessibleInterval} must have at
+ * least D components.
+ * <p>
+ * The deformation field should be interpreted as a d-dimensional
+ * vector field.  A source point is displaced by adding the vector
+ * at that point the the source point's position. 
+ * 
+ * @author John Bogovic &lt;bogovicj@janelia.hhmi.org&gt;
+ *
+ */
+public class DeformationFieldTransform<T extends RealType<T>> implements RealTransform
+{
+
+	private final RealRandomAccessible< T > defFieldReal;
+
+	private final int numDim;
+
+	public DeformationFieldTransform( RandomAccessibleInterval< T > def )
+	{
+        this( Views.interpolate( def, new NLinearInterpolatorFactory< T >() ));
+	}
+
+	public DeformationFieldTransform( RealRandomAccessible< T > defFieldReal )
+	{
+		this.defFieldReal = defFieldReal;
+		this.numDim = defFieldReal.numDimensions() - 1;
+	}
+
+	@Override
+	public int numSourceDimensions()
+	{
+		return numDim;
+	}
+
+	@Override
+	public int numTargetDimensions()
+	{
+		return numDim;
+	}
+
+	@Override
+	public void apply( double[] source, double[] target )
+	{
+		RealRandomAccess< T > defFieldAccess = defFieldReal.realRandomAccess();
```

Comment:
That creates a new `RealRandomAccess` per coordinate transfer which I believe will have significant runtime cost.  I suggest to have one `RealRandomAccess` per transform.  `RealTransform`s are not expected to be used by multiple threads in parallel.